File: src/readers/gbfs/schemaV1/stationInformation.ts
/**
 * # GBFS Station Information Schema V1.1 OR GBFS Station Information Schema V1.0
 * List of all stations, their capacities, and locations. REQUIRED for systems utilizing docks.
 *
 * ## Links
 * - [GBFS Specification V1.1](https://github.com/MobilityData/gbfs/blob/v1.1/gbfs.md#station_informationjson)
 * - [GBFS Specification V1.0](https://github.com/MobilityData/gbfs/blob/v1.0/gbfs.md#station_informationjson)
 */
export type GBFSStationInformationV1 = GBFSStationInformationV11 | GBFSStationInformationV10;

/**
 * # GBFS Station Information Schema V1.1
 * List of all stations, their capacities, and locations. REQUIRED for systems utilizing docks.
 *
 * ## Links
 * - [GBFS Specification V1.1](https://github.com/MobilityData/gbfs/blob/v1.1/gbfs.md#station_informationjson)
 */
export const gbfsStationInformationSchemaV11 = {
  $schema: 'http://json-schema.org/draft-07/schema',
  $id: 'https://github.com/MobilityData/gbfs/blob/v1.1/gbfs.md#station_informationjson',
  description:
    'List of all stations, their capacities, and locations. REQUIRED for systems utilizing docks.',
  type: 'object',
  properties: {
    last_updated: {
      description: 'Last time the data in the feed was updated in POSIX time.',
      type: 'integer',
      minimum: 1450155600,
    },
    ttl: {
      description:
        'Number of seconds before the data in the feed will be updated again (0 if the data should always be refreshed).',
      type: 'integer',
      minimum: 0,
    },
    version: {
      description:
        'GBFS version number to which the feed conforms, according to the versioning framework (added in v1.1).',
      type: 'string',
      const: '1.1',
    },
    data: {
      description: 'Array that contains one object per station as defined below.',
      type: 'object',
      properties: {
        stations: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              station_id: { description: 'Identifier of a station.', type: 'string' },
              name: { description: 'Public name of the station.', type: 'string' },
              short_name: {
                description: 'Short name or other type of identifier.',
                type: 'string',
              },
              lat: {
                description: 'The latitude of the station.',
                type: 'number',
                minimum: -90,
                maximum: 90,
              },
              lon: {
                description: 'The longitude of the station.',
                type: 'number',
                minimum: -180,
                maximum: 180,
              },
              address: {
                description: 'Address where the station is located.',
                type: 'string',
              },
              cross_street: {
                description: 'Cross street or landmark where the station is located.',
                type: 'string',
              },
              region_id: {
                description: 'Identifier of the region where the station is located.',
                type: 'string',
              },
              post_code: {
                description: 'Postal code where the station is located.',
                type: 'string',
              },
              rental_methods: {
                description: 'Payment methods accepted at this station.',
                type: 'array',
                items: {
                  type: 'string',
                  enum: [
                    'KEY',
                    'CREDITCARD',
                    'PAYPASS',
                    'APPLEPAY',
                    'ANDROIDPAY',
                    'TRANSITCARD',
                    'ACCOUNTNUMBER',
                    'PHONE',
                  ],
                },
                minItems: 1,
              },
              capacity: {
                description:
                  'Number of total docking points installed at this station, both available and unavailable.',
                type: 'integer',
                minimum: 0,
              },
              rental_uris: {
                description:
                  'Contains rental uris for Android, iOS, and web in the android, ios, and web fields (added in v1.1).',
                type: 'object',
                properties: {
                  android: {
                    description: 'URI for Android app intent.',
                    type: 'string',
                    format: 'uri',
                  },
                  ios: {
                    description: 'URI for iOS rental app.',
                    type: 'string',
                    format: 'uri',
                  },
                  web: {
                    description: 'Web URL for renting a vehicle at this station.',
                    type: 'string',
                    format: 'uri',
                  },
                },
              },
            },
            required: ['station_id', 'name', 'lat', 'lon'],
          },
        },
      },
      required: ['stations'],
    },
  },
  required: ['last_updated', 'ttl', 'version', 'data'],
};

/**
 * GBFS Station Information Schema V1.1 Interface
 */
export interface GBFSStationInformationV11 {
  /** Last time the data in the feed was updated in POSIX time. */
  last_updated: number;
  /** Number of seconds before the data in the feed will be updated again. */
  ttl: number;
  /** GBFS version number (1.1). */
  version: '1.1';
  /** Data containing an array of stations. */
  data: {
    stations: Array<{
      station_id: string;
      name: string;
      short_name?: string;
      lat: number;
      lon: number;
      address?: string;
      cross_street?: string;
      region_id?: string;
      post_code?: string;
      rental_methods?: Array<
        | 'KEY'
        | 'CREDITCARD'
        | 'PAYPASS'
        | 'APPLEPAY'
        | 'ANDROIDPAY'
        | 'TRANSITCARD'
        | 'ACCOUNTNUMBER'
        | 'PHONE'
      >;
      capacity?: number;
      rental_uris?: {
        android?: string;
        ios?: string;
        web?: string;
      };
    }>;
  };
}

/**
 * # GBFS Station Information Schema V1.0
 * List of all stations, their capacities, and locations. REQUIRED for systems utilizing docks.
 *
 * ## Links
 * - [GBFS Specification V1.0](https://github.com/MobilityData/gbfs/blob/v1.0/gbfs.md#station_informationjson)
 */
export const gbfsStationInformationSchemaV10 = {
  $schema: 'http://json-schema.org/draft-07/schema',
  $id: 'https://github.com/MobilityData/gbfs/blob/v1.0/gbfs.md#station_informationjson',
  description:
    'List of all stations, their capacities, and locations. REQUIRED for systems utilizing docks.',
  type: 'object',
  properties: {
    last_updated: {
      description: 'Last time the data in the feed was updated in POSIX time.',
      type: 'integer',
      minimum: 0,
      maximum: 1924988399,
    },
    ttl: {
      description:
        'Number of seconds before the data in the feed will be updated again (0 if the data should always be refreshed).',
      type: 'integer',
      minimum: 0,
    },
    data: {
      description: 'Array that contains one object per station as defined below.',
      type: 'object',
      properties: {
        stations: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              station_id: { description: 'Identifier of a station.', type: 'string' },
              name: { description: 'Public name of the station.', type: 'string' },
              short_name: {
                description: 'Short name or other type of identifier.',
                type: 'string',
              },
              lat: {
                description: 'The latitude of the station.',
                type: 'number',
                minimum: -90,
                maximum: 90,
              },
              lon: {
                description: 'The longitude of the station.',
                type: 'number',
                minimum: -180,
                maximum: 180,
              },
              address: {
                description: 'Address where the station is located.',
                type: 'string',
              },
              cross_street: {
                description: 'Cross street or landmark where the station is located.',
                type: 'string',
              },
              region_id: {
                description: 'Identifier of the region where the station is located.',
                type: 'string',
              },
              post_code: {
                description: 'Postal code where the station is located.',
                type: 'string',
              },
              rental_methods: {
                description: 'Payment methods accepted at this station.',
                type: 'array',
                items: {
                  type: 'string',
                  enum: [
                    'KEY',
                    'CREDITCARD',
                    'PAYPASS',
                    'APPLEPAY',
                    'ANDROIDPAY',
                    'TRANSITCARD',
                    'ACCOUNTNUMBER',
                    'PHONE',
                  ],
                },
              },
              capacity: {
                description:
                  'Number of total docking points installed at this station, both available and unavailable.',
                type: 'integer',
                minimum: 0,
              },
            },
            required: ['station_id', 'name', 'lat', 'lon'],
          },
        },
      },
      required: ['stations'],
    },
  },
  required: ['last_updated', 'ttl', 'data'],
};

/**
 * GBFS Station Information Schema V1.0 Interface
 */
export interface GBFSStationInformationV10 {
  /** Last time the data in the feed was updated in POSIX time. */
  last_updated: number;
  /** Number of seconds before the data in the feed will be updated again. */
  ttl: number;
  /** Data containing an array of stations. */
  data: {
    stations: Array<{
      station_id: string;
      name: string;
      short_name?: string;
      lat: number;
      lon: number;
      address?: string;
      cross_street?: string;
      region_id?: string;
      post_code?: string;
      rental_methods?: Array<
        | 'KEY'
        | 'CREDITCARD'
        | 'PAYPASS'
        | 'APPLEPAY'
        | 'ANDROIDPAY'
        | 'TRANSITCARD'
        | 'ACCOUNTNUMBER'
        | 'PHONE'
      >;
      capacity?: number;
    }>;
  };
}

File: rust/readers/gbfs/schema_v3/station_information.rs
use crate::readers::{GBFSName, GBFSRentalUri};
use alloc::{string::String, vec::Vec};
use s2json::Geometry;
use serde::{Deserialize, Serialize};

/// # GBFS Station Information Schema V3.1-RC & V3.0
/// List of all stations, their capacities, and locations. REQUIRED for systems utilizing docks.
///
/// ## Links
/// - [GBFS Specification V3.1-RC](https://github.com/MobilityData/gbfs/blob/v3.1-RC/gbfs.md#station_informationjson)
/// - [GBFS Specification V3.0](https://github.com/MobilityData/gbfs/blob/v3.0/gbfs.md#station_informationjson)
pub type GBFSStationInformationV3 = GBFSStationInformationV30;

/// GBFS Station Information Rental Method
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub enum GBFSStationInformationRentalMethodV30 {
    /// Key
    #[default]
    #[serde(rename = "key")]
    Key,
    /// Credit Card
    #[serde(rename = "creditcard")]
    CreditCard,
    /// PayPass
    #[serde(rename = "paypass")]
    PayPass,
    /// Apple Pay
    #[serde(rename = "applepay")]
    ApplePay,
    /// Android Pay
    #[serde(rename = "androidpay")]
    AndroidPay,
    /// Google Wallet
    #[serde(rename = "transitcard")]
    TransitCard,
    /// Cash
    #[serde(rename = "accountnumber")]
    AccountNumber,
    /// Phone
    #[serde(rename = "phone")]
    Phone,
}

/// GBFS Station Information Parking Type
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub enum GBFSStationInformationParkingTypeV30 {
    /// Parking Lot
    #[default]
    #[serde(rename = "parking_lot")]
    ParkingLot,
    /// Street Parking
    #[serde(rename = "street_parking")]
    StreetParking,
    /// Underground Parking
    #[serde(rename = "underground_parking")]
    UndergroundParking,
    /// Sidewalk Parking
    #[serde(rename = "sidewalk_parking")]
    SidewalkParking,
    /// Other
    #[serde(rename = "other")]
    Other,
}

/// GBFS Station Information Name
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSStationV3VehicleCapacity {
    /// List of vehicle types.
    pub vehicle_type_ids: Vec<String>,
    /// Number of vehicles.
    pub count: u64,
}

/// Information about a single station.
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSStationV3 {
    /// Identifier of the station.
    pub station_id: String,
    /// Public name of the station.
    pub name: Vec<GBFSName>,
    /// The latitude of the station.
    /// **Minimum**: -90
    /// **Maximum**: 90
    pub lat: f64,
    /// The longitude of the station.
    /// **Minimum**: -180
    /// **Maximum**: 180
    pub lon: f64,
    /// Short name or alternative identifier for the station.
    pub short_name: Option<Vec<GBFSName>>,
    /// Address where the station is located.
    pub address: Option<String>,
    /// Cross street or landmark where the station is located.
    pub cross_street: Option<String>,
    /// Identifier of the region where the station is located.
    pub region_id: Option<String>,
    /// Postal code where the station is located.
    pub post_code: Option<String>,
    /// Hours of operation for the station in OSM opening_hours format.
    pub station_opening_hours: Option<String>,
    /// Payment methods accepted at the station.
    /// **Enum**: ['key', 'creditcard', 'paypass', 'applepay', 'androidpay', 'transitcard', 'accountnumber', 'phone']
    pub rental_methods: Option<Vec<GBFSStationInformationRentalMethodV30>>,
    /// Is this station a location with or without physical infrastructure? (added in v2.1-RC)
    pub is_virtual_station: Option<bool>,
    /// A multipolygon describing the area of a virtual station. (added in v2.1-RC)
    pub station_area: Option<Geometry>,
    /// Type of parking station. (added in v2.3)
    /// **Enum**: ['parking_lot', 'street_parking', 'underground_parking', 'sidewalk_parking', 'other']
    pub parking_type: Option<GBFSStationInformationParkingTypeV30>,
    /// Are parking hoops present at this station? (added in v2.3)
    pub parking_hoop: Option<bool>,
    /// Contact phone of the station. (added in v2.3)
    pub contact_phone: Option<String>,
    /// Total docking points installed at the station, both available and unavailable.
    /// **Minimum**: 0
    pub capacity: Option<u64>,
    /// Parking capacity for virtual stations per vehicle type.
    pub vehicle_types_capacity: Option<Vec<GBFSStationV3VehicleCapacity>>,
    /// Docking capacity per vehicle type at the station.
    pub vehicle_docks_capacity: Option<Vec<GBFSStationV3VehicleCapacity>>,
    /// Are valet services provided at the station? (added in v2.1-RC)
    pub is_valet_station: Option<bool>,
    /// Does the station support charging of electric vehicles? (added in v2.3-RC)
    pub is_charging_station: Option<bool>,
    /// Rental URIs for Android, iOS, and web.
    pub rental_uris: Option<GBFSRentalUri>,
}

/// Data object containing station information.
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSStationInformationV30Data {
    /// List of stations with their attributes.
    pub stations: Vec<GBFSStationV3>,
}

/// # GBFS Station Information Schema V3.0
/// List of all stations, their capacities, and locations. REQUIRED for systems utilizing docks.
///
/// ## Links
/// - [GBFS Specification](https://github.com/MobilityData/gbfs/blob/v3.0/gbfs.md#station_informationjson)
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct GBFSStationInformationV30 {
    /// Last time the data in the feed was updated in RFC3339 format.
    /// **Format**: date-time
    pub last_updated: String,
    /// Number of seconds before the data in the feed will be updated again (0 if the data should always be refreshed).
    /// **Minimum**: 0
    pub ttl: u64,
    /// GBFS version number to which the feed conforms.
    /// **Const**: '3.0'
    pub version: String,
    /// Data object containing station information.
    pub data: GBFSStationInformationV30Data,
}
